packages/lab/src/carousel/CarouselContext.tsx

```diff
@@ -0,0 +1,154 @@
+  children,
+  activeSlideIndex = 0,
+  visibleSlides = 1,
+  id,
+}: {
+  children: ReactNode;
+  activeSlideIndex?: number;
+  visibleSlides?: number;
+  id?: string;
+}) {
+  const [firstVisibleSlide, setFirstVisibleSlide] = useState(activeSlideIndex);
+  const [visibleFocus, setVisibleFocus] = useState(0);
+  const [sliderW, setSliderW] = useState(0);
+  const slides = useRef<Map<number, HTMLDivElement>>(new Map());
+  const [slidesCount, setSlidesCount] = useState(slides.current.size);
+  useEffect(() => {
+    if (containerRef.current) {
+      scrollToSlide(activeSlideIndex);
+    }
+  }, [activeSlideIndex]);
+
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  const registerSlide = useCallback((element: HTMLDivElement) => {
+    const assignedIndex = slides.current.size;
+    slides.current.set(assignedIndex, element);
+    setSlidesCount(assignedIndex + 1);
+    return assignedIndex;
+  }, []);
+
+  const unregisterSlide = useCallback((index: number) => {
+    slides.current.delete(index);
+    setSlidesCount(slides.current.size);
+  }, []);
+
+  const updateActiveFromScroll = (scrollLeft: number) => {
+    const newIndex = Math.round(scrollLeft / (sliderW / visibleSlides)) || 0;
+    if (newIndex !== firstVisibleSlide) {
+      setFirstVisibleSlide(newIndex);
+    }
+  };
+
+  const handleResize = useCallback(() => {
+    if (!containerRef.current) return;
+    setSliderW(containerRef.current.offsetWidth);
+  }, []);
+
+  useResizeObserver({ ref: containerRef, onResize: handleResize });
+
+  const scrollToSlide = (index: number) => {
+    if (!containerRef.current) return;
+    const sliderWidth = containerRef.current.offsetWidth;
+    const slideWidth = sliderWidth / visibleSlides;
+    const targetScrollLeft = index * slideWidth;
+    containerRef.current.scrollTo({
+      left: targetScrollLeft,
+      behavior: "smooth",
+    });
+  };
+
+  const focusSlide = (index: number) => {
+    slides.current.get(index)?.focus();
+  };
+
+  const nextSlide = (event: MouseEvent | KeyboardEvent) => {
+    const nextSlide = firstVisibleSlide + 1;
+    if (!containerRef.current || nextSlide >= slidesCount) return;
+    if (event.type !== "click") {
+      focusSlide(nextSlide);
+      if (visibleFocus < visibleSlides - 1) {
+        setVisibleFocus((prev) => prev + 1);
+      }
+      return;
+    }
+    scrollToSlide(nextSlide);
+  };
+
+  const prevSlide = (event: MouseEvent | KeyboardEvent) => {
+    const previousSlide = firstVisibleSlide - 1;
+    if (!containerRef.current || firstVisibleSlide < 0) return;
+    if (event.type !== "click") {
+      if (visibleFocus >= 0) {
+        focusSlide(previousSlide + visibleFocus);
+        if (previousSlide === firstVisibleSlide) {
+          setVisibleFocus((prev) => prev - 1);
+        }
+        return;
+      }
+      focusSlide(previousSlide);
+    }
+    scrollToSlide(previousSlide);
+  };
+
+  return (
+    <CarouselContext.Provider
+      value={{
+        visibleSlides,
```

Comment:
in date picker we have state and helpers.
is this a pattern that would make the API easier to understand?